client/js/webpush.js

```diff
@@ -0,0 +1,129 @@
+const storage = require("./localStorage");
+const socket = require("./socket");
+
+const pushNotificationsButton = $("#pushNotifications");
+let clientSubscribed = null;
+let applicationServerKey;
+
+module.exports.configurePushNotifications = (subscribedOnServer, key) => {
+	applicationServerKey = key;
+
+	// If client has push registration but the server knows nothing about it,
+	// this subscription is broken and client has to register again
+	if (clientSubscribed === true && subscribedOnServer === false) {
+		pushNotificationsButton.attr("disabled", true);
+
+		navigator.serviceWorker.register("service-worker.js").then((registration) => registration.pushManager.getSubscription().then((existingSubscription) => {
+			if (!existingSubscription) {
+				return;
+			}
+
+			return existingSubscription.unsubscribe().then((successful) => {
+				if (successful) {
+					alternatePushButton().removeAttr("disabled");
+				}
+			});
+		}));
+	}
+};
+
+if (isAllowedServiceWorkersHost()) {
+	$("#pushNotificationsHttps").hide();
+
+	if ("serviceWorker" in navigator) {
+		navigator.serviceWorker.register("service-worker.js").then((registration) => {
+			if (!registration.pushManager) {
+				return;
+			}
+
+			return registration.pushManager.getSubscription().then((subscription) => {
+				$("#pushNotificationsUnsupported").hide();
+
+				pushNotificationsButton
+					.removeAttr("disabled")
+					.on("click", onPushButton);
+
+				clientSubscribed = !!subscription;
+
+				if (clientSubscribed) {
+					alternatePushButton();
+				}
+			});
+		}).catch((err) => {
+			$("#pushNotificationsUnsupported p").text(err);
+		});
+	}
+}
+
+function onPushButton() {
+	pushNotificationsButton.attr("disabled", true);
+
+	navigator.serviceWorker.register("service-worker.js").then((registration) => registration.pushManager.getSubscription().then((existingSubscription) => {
+		if (existingSubscription) {
+			socket.emit("push:unregister");
+
+			return existingSubscription.unsubscribe().then((successful) => {
+				if (successful) {
+					alternatePushButton().removeAttr("disabled");
+				}
+			});
+		}
+
+		return registration.pushManager.subscribe({
+			applicationServerKey: urlBase64ToUint8Array(applicationServerKey),
+			userVisibleOnly: true
+		}).then((subscription) => {
+			const rawKey = subscription.getKey ? subscription.getKey("p256dh") : "";
+			const key = rawKey ? window.btoa(String.fromCharCode.apply(null, new Uint8Array(rawKey))) : "";
+			const rawAuthSecret = subscription.getKey ? subscription.getKey("auth") : "";
+			const authSecret = rawAuthSecret ? window.btoa(String.fromCharCode.apply(null, new Uint8Array(rawAuthSecret))) : "";
+
+			socket.emit("push:register", {
+				token: storage.get("token"),
+				endpoint: subscription.endpoint,
+				keys: {
+					p256dh: key,
+					auth: authSecret
+				}
+			});
+
+			alternatePushButton().removeAttr("disabled");
+		});
+	})).catch((err) => {
+		console.error(err);
+		window.alert(err);
```

Comment:
Can we add an extra block next to `pushNotificationWarning` stuff? We really can't be reporting in `window.alert`...